hippo.go

```diff
@@ -126,80 +143,128 @@ func (c *Client) Do(ctx context.Context, req *http.Request) ([]byte, error) {
+		}
+		responseBytes, err := c.Do(ctx, req)
+		if err != nil {
+			return ret, fmt.Errorf("Error POSTing populators to %s - [%s] - underlying error: %s", url, ret, err)
+		}
+
+		if i == 0 {
+			// first response returns the batch GUID, which we need to include in subsequent batches
+			apiResponse := APIServerResponse{}
+			if err := json.Unmarshal(responseBytes, &apiResponse); err != nil {
+				return ret, fmt.Errorf("Error unmarshalling API batch response - [%s] - underlying error: %s", ret, err)
+			}
+			if apiResponse.GUID == "" {
+				return ret, fmt.Errorf("API response did not include a GUID for subsequent batches - [%s] - underlying error: %s", ret, apiResponse.Error)
+			}
+			ret.BatchGUID = apiResponse.GUID
+		}
 
-	// swap out our local copy with a compacted one that ensures all criteria are grouped by value
-	tmp := compactTagBatchPart(*rFull)
-	tmp.Sender = c.Sender
-	rFull = &tmp
+		// update response
+		ret.PartsSent++
+		ret.UpsertsSent += len(batchPart.Upserts)
+		ret.DeletesSent += len(batchPart.Deletes)
 
+		// slow down the HTTP batches a bit to avoid rate limiting
+		time.Sleep(time.Second)
+	}
+
+	return ret, nil
+}
+
+// split a batch into parts small enough for sending through Kentik API
+func (c *Client) split(rFull *TagBatchPart) ([]TagBatchPart, error) {
 	encode := func(r *TagBatchPart) ([]byte, error) {
 		if b, err := json.Marshal(r); err != nil {
-			return nil, err
+			return nil, fmt.Errorf("Error JSON-marshalling batch: %s", err)
 		} else {
 			return b, nil
 		}
 	}
 
-	// If the size is small enough, just return here. Assume default case
-	init, err := encode(rFull)
+	serializedBytes, err := encode(rFull)
 	if err != nil {
-		return nil, 0, err
-	} else if len(init) < MAX_HIPPO_SIZE {
-		return [][]byte{init}, len(rFull.Upserts), nil
+		return nil, err
+	}
+
+	outgoingRequestSize := c.OutgoingRequestSize
+	if outgoingRequestSize <= 0 {
+		outgoingRequestSize = DEFAULT_MAX_HIPPO_SIZE
+	}
+
+	if len(serializedBytes) < outgoingRequestSize {
+		return []TagBatchPart{*rFull}, nil
 	}
 
 	// Here, have to split this request into a group
-	parts := (len(init) / MAX_HIPPO_SIZE) + 1
+	parts := (len(serializedBytes) / outgoingRequestSize) + 1
 	upsertPerPart := len(rFull.Upserts) / parts
-	reqArrary := make([][]byte, parts)
+	ret := make([]TagBatchPart, parts)
 	lastUp := 0
-	numUpserts := 0
 
 	for i := 0; i < parts-1; i++ {
-		rTmp := &TagBatchPart{
+		ret[i] = TagBatchPart{
 			ReplaceAll: rFull.ReplaceAll,
 			IsComplete: false,
 			TTLMinutes: rFull.TTLMinutes,
 			Upserts:    rFull.Upserts[lastUp : lastUp+upsertPerPart],
-			Sender:     c.Sender,
+			Sender:     rFull.Sender,
 		}
 
-		next, err := encode(rTmp)
-		if err != nil {
-			return nil, 0, err
-		} else {
-			reqArrary[i] = next
-		}
 		lastUp += upsertPerPart
-		numUpserts += len(rTmp.Upserts)
 	}
 
 	// Last one has to be handled seperately
-	rLast := &TagBatchPart{
+	ret[parts-1] = TagBatchPart{
```

Comment:
nit: line above spelling/typo error on `seperately` instead of `separately`

Reply:
pretty sure you're wrong, but i changed it anyway